File: 7-grafos/1100.py
queue,board,init,end=None,None,None,None

def moveKnight(x,y):
    tot=[]
    for i in range(2):
        if x+i+1<9 and y+2-i<9 and not board[y+2-i][x+i+1]:
            tot.append((x+i+1)*10+y+2-i)
            board[y+2-i][x+i+1]=1
        if x+i+1<9 and y-2+i>0 and not board[y-2+i][x+i+1]:
            tot.append((x+i+1)*10+y-2+i)
            board[y-2+i][x+i+1]=1
    for i in range(2):
        if x-i-1>0 and y+2-i<9 and not board[y+2-i][x-i-1]:
            tot.append((x-i-1)*10+y+2-i)
            board[y+2-i][x-i-1]=1
        if x-i-1>0 and y-2+i>0 and not board[y-2+i][x-i-1]:
            tot.append((x-i-1)*10+y-2+i)
            board[y-2+i][x-i-1]=1
    return tot
def bfs(cont,queue):
    aux=[]
    while (len(queue)):
        if(queue[-1]==end):
            print("To get from {}{} to {}{} takes {} knight moves.".format(chr(int((init-init%10)/10+96)),init%10,chr(int((end-end%10)/10+96)),end%10,cont))
            return
        for i in moveKnight(int((queue[-1]-queue[-1]%10)/10),queue[-1]%10):
            aux.append(i)
            #print(chr(int((i-i%10)/10)+96)+str(i%10))
        queue.pop()
    queue=aux.copy()
    aux.clear()
    bfs(cont+1,queue)
while 1:
    try:x=input()
    except:break
    init=(ord(x[0])-96)*10+int(x[1])
    end=(ord(x[3])-96)*10+int(x[4])
    queue=[init]
    board=[[0 for i in range(10)] for j in range(10)]
    board[init%10][int((init-init%10)/10)]=1
    bfs(0,queue)

File: 7-grafos/1100.cpp
#include<iostream>
#include<vector>
#include<queue>
//
int**board,end,init;
std::queue<int>fila;

std::vector<int> actionHorse(int x,int y){
	std::vector<int>store;
	for(int i=0;i<2;++i){
		if(x+i+1<8&&y+2-i<8&&!board[y+2-i][x+i+1]){
			store.push_back(x+i+1+(y+2-i)*10);
			board[y+2-i][x+i+1]=1;
		}
		if(x+i+1<8&&y-2+i>-1&&!board[y-2+i][x+i+1]){
			store.push_back(x+i+1+(y-2+i)*10);
			board[y-2+i][x+i+1]=1;
		}
	}
	for(int i=0;i<2;++i){
		if(x-i-1>-1&&y+2-i<8&&!board[y+2-i][x-i-1]){
			store.push_back(x-i-1+(y+2-i)*10);
			board[y+2-i][x-i-1]=1;
		}
		if(x-i-1>-1&&y-2+i>-1&&!board[y-2+i][x-i-1]){
			store.push_back(x-i-1+(y-2+i)*10);
			board[y-2+i][x-i-1]=1;
		}
	}
	return store;
}

void bfs(int cont){
	std::queue<int>aux;
	for(;!fila.empty();){
		int x=fila.front()%10,y=(fila.front()-fila.front()%10)/10;
		//std::cout<<fila.front()<<"|";
		if(fila.front()==end){
			std::cout<<"To get from "<<char((init%10)+97)<<(init-init%10)/10+1<<" to "<<char((end%10)+97)<<(end-end%10)/10+1<<" takes "<<cont<<" knight moves.\n";
			fila=std::queue<int>();
			return;
		}
		for(int analyse:actionHorse(x,y)){
			int auxX=analyse%10,auxY=(analyse-analyse%10)/10;
			//std::cout<<char(auxX+97)<<auxY<<" "<<" ";
			aux.push(analyse);
		}
		fila.pop();
	}
	fila=aux;
	bfs(cont+1);
}
//
int main(){
	std::string inp;
	for(;std::getline(std::cin,inp)&&inp.length();){
		init=(int)inp[0]-97+(inp[1]-'0'-1)*10;
		end=((int)inp[3]-97)+(inp[4]-'0'-1)*10;
		board=(int**)malloc(8*sizeof(int*));
		for(int i=0;i<8;++i)board[i]=(int*)calloc(8,sizeof(int));
		board[(init-init%10)/10][init%10]=1;
		fila.push(init);
		bfs(0);
	}
}
